src/audio/module_adapter/module_adapter_ipc3.c

```diff
@@ -0,0 +1,330 @@
+			buffer_stream_writeback(sink_c, mod->output_buffers[i].size);
+		if (mod->output_buffers[i].size)
+			comp_update_buffer_produce(sink_c, mod->output_buffers[i].size);
+	}
+
+	mod->total_data_produced += mod->output_buffers[0].size;
+
+	/* release all sink buffers */
+	for (i = 0; i < num_output_buffers; i++) {
+		buffer_release(sinks_c[i]);
+		mod->output_buffers[i].size = 0;
+	}
+
+	return 0;
+out:
+	for (i = 0; i < num_output_buffers; i++) {
+		buffer_release(sinks_c[i]);
+		mod->output_buffers[i].size = 0;
+	}
+
+	for (i = 0; i < num_input_buffers; i++) {
+		buffer_release(source_c[i]);
+		mod->input_buffers[i].size = 0;
+		mod->input_buffers[i].consumed = 0;
+	}
+
+	return ret;
+}
+
+static int module_source_status_count(struct comp_dev *dev, uint32_t status)
+{
+	struct list_item *blist;
+	int count = 0;
+
+	/* count source with state == status */
+	list_for_item(blist, &dev->bsource_list) {
+		/*
+		 * FIXME: this is racy, state can be changed by another core.
+		 * This is implicitly protected by serialised IPCs. Even when
+		 * IPCs are processed in the pipeline thread, the next IPC will
+		 * not be sent until the thread has processed and replied to the
+		 * current one.
+		 */
+		struct comp_buffer *source = container_of(blist, struct comp_buffer,
+							  sink_list);
+		struct comp_buffer __sparse_cache *source_c = buffer_acquire(source);
+
+		if (source_c->source && source_c->source->state == status)
+			count++;
+		buffer_release(source_c);
+	}
+
+	return count;
+}
+
+int module_adapter_set_state(struct processing_module *mod, struct comp_dev *dev,
+			     int cmd)
+{
+	if (mod->num_input_buffers > 1) {
+		bool sources_active;
+		int ret;
+
+		sources_active = module_source_status_count(dev, COMP_STATE_ACTIVE) ||
+				 module_source_status_count(dev, COMP_STATE_PAUSED);
+
+		/* don't stop/start module if one of the sources is active/paused */
+		if ((cmd == COMP_TRIGGER_STOP || cmd == COMP_TRIGGER_PRE_START) && sources_active) {
+			dev->state = COMP_STATE_ACTIVE;
+			return PPL_STATUS_PATH_STOP;
+		}
+
+		ret = comp_set_state(dev, cmd);
+		if (ret == COMP_STATUS_STATE_ALREADY_SET)
+			return PPL_STATUS_PATH_STOP;
+
+		return ret;
+	}
+
+	return comp_set_state(dev, cmd);
+}
+
+int module_adapter_get_attribute(struct comp_dev *dev, uint32_t type, void *value)
+{
+	return -EINVAL;
+}
+
+int module_set_large_config(struct comp_dev *dev, uint32_t param_id, bool first_block,
+			    bool last_block, uint32_t data_offset, const char *data)
+{
+	return 0;
+}
+
+int module_get_large_config(struct comp_dev *dev, uint32_t param_id, bool first_block,
+			    bool last_block, uint32_t *data_offset, char *data)
+{
+	return 0;
+}
+
+int module_adapter_bind(struct comp_dev *dev, void *data)
+{
+	return 0;
+}
+
+int module_adapter_unbind(struct comp_dev *dev, void *data)
+{
+	return 0;
+}
+
+uint64_t module_adapter_get_total_data_processed(struct comp_dev *dev,
+						 uint32_t stream_no, bool input)
+{
+	return 0;
+}
+
```

Comment:
These are IPC4 only APIs, I would move them to the common header and make then static inline if using IPC3. That way they all become NOPs for all IPC3 users (and CC will eliminate).

Reply:
yes, however, if move to generic header file:
1. need add #if ipc3
2. ipc4  implementation have more than 200 lines, not suitable to move to header. 
    still need to keep in ipc4 source file.
3. somehow, with current implementation, code looks symmetry.

Finally, those should be empty for ipc3 build as long as with -o3, each compiler should eliminate those empty function.

Reply:
> 

Thats fine having a single ```#if IPC3``` in the header, this also make sure we are not building any code that is not needed even in debug. Fwiw, this is what Linux does too.